lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -2744,7 +2744,50 @@ void LocalAllocOp::getEffects(
                        mlir::triton::gpu::SharedMemory::get());
 }
 
-//===----------------------------------------------------------------------===//
+LogicalResult MemDescSubviewOp::verify() {
+  auto srcTy = getDesc().getType();
+  auto dstTy = getType();
+
+  if (srcTy.getElementType() != dstTy.getElementType()) {
+    return emitError("result element type must match desc element type");
+  }
+  if (getOffsets().size() != srcTy.getRank()) {
+    return emitError("offsets must have the same rank as input");
+  }
+  if (srcTy.getRank() < dstTy.getRank()) {
+    return emitError("result rank must be less than or equal to input rank");
+  }
+  auto rankDiff = srcTy.getRank() - dstTy.getRank();
+  for (int i = 0; i < dstTy.getRank(); i++) {
+    if (dstTy.getDimSize(i) > srcTy.getDimSize(i + rankDiff)) {
+      return emitError(
+                 "result shape cannot be larger than input shape at dimension ")
+             << i;
+    }
+  }
+
+  auto srcEnc = srcTy.getEncoding();
+  auto dstEnc = dstTy.getEncoding();
+  if (!!srcEnc != !!dstEnc) {
```

Comment:
what does `!!` do here? How is that different than `(srcEnc == dstEnc)`?

Reply:
!! converts to a boolean.  That's different than checking that the encodings are the same.